tensorflow/compiler/mlir/xla/transforms/hlo_outline_fusion.cc

```diff
@@ -0,0 +1,553 @@
+  SmallVector<Value, 4> inputs;
+  DenseSet<Value> input_set;
+  std::unordered_set<Operation*> op_set(pattern.begin(), pattern.end());
+
+  for (auto& op : pattern) {
+    for (int i = 0; i < op->getNumOperands(); ++i) {
+      auto operand_op = op->getOperand(i).getDefiningOp();
+      if (op_set.find(operand_op) != op_set.end()) {
+        // skip if defining op is in the pattern
+        continue;
+      }
+      auto value = op->getOperand(i);
+      if (input_set.insert(value).second) {
+        inputs.push_back(value);
+      }
+    }
+  }
+  return std::move(inputs);
+}
+
+SmallVector<Value, 4> GetOutputsOfFusionPattern(const FusionPattern& pattern) {
+  SmallVector<Value, 4> outputs;
+  DenseSet<Value> output_set;
+  std::unordered_set<Operation*> op_set(pattern.begin(), pattern.end());
+  for (auto& op : pattern) {
+    for (int i = 0; i < op->getNumResults(); ++i) {
+      Value result = op->getResult(i);
+      bool has_external_user = false;
+      for (auto iter = result.use_begin(), end = result.use_end(); iter != end; ++iter) {
+        if (op_set.find(iter->getOwner()) == op_set.end()) {
+          has_external_user = true;
+          break;
+        }
+      }
+      if (has_external_user) {
+        if (output_set.insert(result).second) {
+          outputs.push_back(result);
+        }
+      }
+    }
+  }
+  return outputs;
+}
+
+SmallVector<Value, 4> GetInputusAndOutputsOfOp(Operation* op) {
+  SmallVector<Value, 4> ios;
+  ios.reserve(op->getNumOperands() + op->getNumResults());
+  for (int idx = 0; idx < op->getNumOperands() ; ++idx) {
+    ios.push_back(op->getOperand(idx));
+  }
+  for (int idx = 0; idx < op->getNumResults(); ++idx) {
+    ios.push_back(op->getResult(idx));
+  }
+  return ios;
+}
+
+FusionPattern MergeFusionPattern(
+    const FusionPattern& lhs, const FusionPattern& rhs) {
+  FusionPattern pattern(lhs);
+  pattern.insert(pattern.end(), rhs.begin(), rhs.end());
+  return pattern;
+}
+
+class ShapeConstraintAnalysis {
+ public:
+   ShapeConstraintAnalysis(const SmallVector<Operation*, 4>& op_list)
+       : op_list_(op_list) {
+     InitState();
+     PropagateEquality();
+   }
+
+   bool HasSameShape(Value lhs, Value rhs) {
+     return (shape_checker_.count(lhs) &&
+         shape_checker_.count(rhs) &&
+         shape_checker_[lhs].Get() == shape_checker_[rhs].Get());
+   }
+
+ private:
+  void InitState() {
+    for (auto op : op_list_) {
+      for (auto v : GetInputusAndOutputsOfOp(op)) {
+        shape_checker_[v].Get() = v;
+      }
+    }
+  }
+
+  void PropagateEquality() {
+    bool converged = true;
+    do {
+      converged = true;
+      for (auto op : op_list_) {
+        if (IsElementWise(op)) {
+          auto ios = GetInputusAndOutputsOfOp(op);
+          for (auto v : ios) {
+            Value ref = ios[0];
+            if (shape_checker_[ref].Get() != shape_checker_[v].Get()) {
+              converged = false;
+              shape_checker_[ref].Merge(&shape_checker_[v]);
```

Comment:
Is this Merge() action safe or not?
For simple pattern, it is expected to be safe.
For patterns which are more complicated, such as there is a reduction operation in the middle(although I think for kInput/kLoop this will not happen), the rule assumed here will be broken.